### PR TITLE
[config-plugins] Fixed bug on mergeContents when tags have the same prefix

### PR DESCRIPTION
**Pull request by alfonsocj**
# Why
- Running multiple `mergeContents` on the same file **with tags that start the same way** (e.g. `some-tag` and `some-tag-test`) could remove non-intended generated code.

For example:
```
function addCodeOnCreateBeforeSuper(src: string) {
  return mergeContents({
    tag: 'on-create-before-super',
    src,
    newSrc,
    anchor: /super\.onCreate\(\);/,
    offset: 0,
    comment: '//',
  });
}

function addCodeOnCreate(src: string) {
  return mergeContents({
    tag: 'on-create',
    src,
    newSrc: 'some code after super',
    anchor: /super\.onCreate\(\);/,
    offset: 1,
    comment: '//',
  });
}

function withSomePlugin() {
  return withMainApplication(expoConfig, (config) => {
    config.modResults.contents = addCodeOnCreateBeforeSuper(config.modResults.contents);
    config.modResults.contents = addCodeOnCreate(config.modResults.contents);
  });
}
```

This will cause the added code from `addCodeOnCreateBeforeSuper` to be removed.

# How
- `removeGeneratedContents` uses `getGeneratedSectionIndexes` to get the tag indices.
- `getGeneratedSectionIndexes` way to identify sections was not taking into account that tags on the same file could start in the same way. Using regex to identify if the generated comment ends with the exact tag solves issue. 

# Test Plan

- Added a new test to identify the issue and prevent regressions.

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/@expo/config-plugins/src/utils/generateCode.ts
Comment (alfonsocj) [inline packages/@expo/config-plugins/src/utils/generateCode.ts:14]: replaced `string.includes` to be able to detect if the comment ends with the right tag.

File: packages/@expo/config-plugins/src/utils/generateCode.ts
Comment (alfonsocj) [inline packages/@expo/config-plugins/src/utils/generateCode.ts:13]: This one I didn't need to replace `line.includes`, I just wanted to be consistent with the next line.

File: packages/@expo/config-plugins/src/utils/generateCode.ts
Comment (EvanBacon) [inline packages/@expo/config-plugins/src/utils/generateCode.ts:14]:
```suggestion
  const end = contents.findIndex((line) => new RegExp(`@generated end ${tag}$`).test(line));
```

Please use `new` to construct the object.

**Review by EvanBacon — APPROVED** [review]
Small xit, then add changelog, and lgtm

File: packages/@expo/config-plugins/CHANGELOG.md
Comment (byCedric) [inline packages/@expo/config-plugins/CHANGELOG.md:13]:
```suggestion
- Fixed bug on mergeContents when tags have the same prefix. ([#25972](https://github.com/expo/expo/pull/25972) by [@alfonsocj](https://github.com/alfonsocj))
- Separate out runtime version setting method for eas-cli. ([#25874](https://github.com/expo/expo/pull/25874) by [@wschurman](https://github.com/wschurman))
```

File: packages/@expo/config-plugins/src/utils/__tests__/generateCode-test.ts
Comment (byCedric) [inline packages/@expo/config-plugins/src/utils/__tests__/generateCode-test.ts:22]: Just a small nit: maybe an inline snapshot could make this a bit more readable, e.g. just how it's supposed to look 😄

File: packages/@expo/config-plugins/src/utils/generateCode.ts
Comment (byCedric) [inline packages/@expo/config-plugins/src/utils/generateCode.ts:13]: I'm a bit hesitant by adding ` -` to this regex, mostly by [this comment](https://github.com/expo/expo/blob/b029de00ba7006db5bfc5f6ce9091229ef0e859c/packages/%40expo/config-plugins/src/utils/generateCode.ts#L119C37-L119C37):
```
// Everything after the `${tag} ` is unversioned and can be freely modified without breaking changes.
return `${comment} @generated begin ${tag} - expo prebuild (DO NOT MODIFY) ${hashKey}`;
```

We could still go with this. But, if that concern is shared, we could also force the `tags` to be slugs to replace spaces with `-`. With that, we can just test if there is no space at the end, but not sure about the current usage (are there a lot of tags without spaces currently being used).

**Review by byCedric — COMMENTED** [review]
Looking good, thanks! Added a nit, but I'm not a code owner, so I'll defer to @brentvatne or @EvanBacon.

**Review by byCedric — COMMENTED** [review]
Looking good, thanks! Added a few nits, but I'm not a code owner, so I'll defer to @brentvatne or @EvanBacon.

File: packages/@expo/config-plugins/src/utils/generateCode.ts
Comment (byCedric) [reply packages/@expo/config-plugins/src/utils/generateCode.ts:13]: I see evan already commented, closing this.

File: packages/@expo/config-plugins/src/utils/__tests__/generateCode-test.ts
Comment (alfonsocj) [inline packages/@expo/config-plugins/src/utils/__tests__/generateCode-test.ts:31]: @byCedric Should be more readable now! Thanks for the suggestion.

**Pull request merged**
